### PR TITLE
Parameterize App with a (monad) m

### LINKED ISSUE #18: Parameterize the App type on the base monad.
Currently, the application monad transformer stack is stuck on `IO` as a base monad. This forces transformers to be layered on top of the `App`, which can be undesirable. Parameterizing the `App` type on the base monad will allow it to transform other monads, like `Identity` (for pure testing), `Conduit` (for streaming), and `ST` (for locally stateful computations that can be purely run).

- [ ] Change the type definition to be `newtype AppT m a = ...`
- [ ] Provide a type synonym for the common case of `AppT IO`.

### PR DESCRIPTION
**Pull request by 0rca**
This hopefully resolves #18

### CODE REVIEW

File: src/Api.hs [hunk 1]
```diff
@@ -29,12 +29,12 @@ userApp cfg = serve (Proxy :: Proxy UserAPI) (appToServer cfg)
 appToServer :: Config -> Server UserAPI
 appToServer cfg = enter (convertApp cfg) userServer
 
--- | This function converts our 'App' monad into the @ExceptT ServantErr
--- IO@ monad that Servant's 'enter' function needs in order to run the
+-- | This function converts our 'AppT m' monad into the @ExceptT ServantErr
```
Comment (parsonsmatt) [inline src/Api.hs]: Haddock syntax is a little weird. Single quoted identifiers are linked, but IIRC only if it's a single word. The right way to write this is `@'AppT' m@` -- `@` denotes a code block, and `'` denotes linking.

File: test/UserDbSpec.hs [hunk 2]
```diff
@@ -19,11 +20,11 @@ import Database.Persist.Types (Filter)
 import Servant
 
 import Api.User
-import Config (App(..), Config(..), Environment(..), makePool)
+import Config (AppT(..), Config(..), Environment(..), makePool)
 import Models
 import qualified Data.Text as T
 
-runAppToIO :: Config -> App a -> IO a
+runAppToIO :: Config -> AppT IO a -> IO a
```
Comment (parsonsmatt) [inline test/UserDbSpec.hs]: We can continue using the synonym `App` for the common case of `AppT IO`.

File: src/Config.hs [hunk 3]
```diff
@@ -27,12 +27,14 @@ import           System.Environment                   (lookupEnv)
 --
 -- By encapsulating the effects in our newtype, we can add layers to the
 -- monad stack without having to modify code that uses the current layout.
-newtype App a
-    = App
-    { runApp :: ReaderT Config (ExceptT ServantErr IO) a
+newtype AppT m a
```
Comment (parsonsmatt) [inline src/Config.hs]: :+1:

File: src/Api.hs (hunk 1 above)
Comment (0rca) [reply src/Api.hs]: That's good to know, thank you.

File: test/UserDbSpec.hs (hunk 2 above)
Comment (0rca) [reply test/UserDbSpec.hs]: Right. Fixed that

**Comment by parsonsmatt** [comment]
Fantastic, thanks for the PR! :smile:

**Comment by 0rca** [comment]
My pleasure 😊

**Pull request merged**